rfcs/0005-fork-selection.md

```diff
@@ -0,0 +1,166 @@
+|    1 |         1 |      16 |                   17 |
+
+The lockout for vote 1 will not increase from 16 until the stack contains 5 votes.
+
+### Slashing and Rewards
+
+The purpose of the lockout is to force a voter to commit opportunity cost to a specific fork.  Voters that violate the lockouts and vote for a diverging fork within the lockout should be punished.  Slashing or simply freezing the voter from rewards for a long period of time can be used as punishment.
+
+Voters should be rewarded for selecting the fork that the rest of the network selected as often as possible.  This is well-aligned with generating a reward when the vote stack is full and the oldest vote needs to be dequeued.  Thus a reward should be generated for each successful dequeue.
+
+### Economic Finality
+
+Economic finality can be calculated as the total opportunity costs due to the vote lockout at a given fork.  As a votes get further and further, buried the economic finality increases because the cost of unrolling would be the total loss of all the reward from the lockouts at that fork.
+
+### Thresholds
+
+Each voter can independently set a threshold of network commitment to a fork before that voter commits to a fork.  For example, at vote stack index 7, the lockout is 256 time units.  A voter may withhold votes and let votes 0-7 expire unless the vote at index 7 has at greater than 50% commitment in the network.  This allows each voter to independently control how much risk to commit to a fork.  Committing to a fork faster would allow the voter to earn more rewards.
+
+### Algorithm parameters
+
+These parameters need to be tuned.
+
+* Number of votes in the stack before dequeue occurs (32).
+* Rate of growth for lockouts in the stack (2x).
+* Starting default lockout (2).
+* Threshold depth for minimum network commitment before committing to the fork (8).
+* Minimum network commitment size at threshold depth (50%+).
+
+### Free Choice
+
+A "Free Choice" is an unenforcible node action.  A node that maximizes self-reward over all possible futures should behave in such a way that the system is stable, and the local greedy choice should result in a greedy choice over all possible futures.  A set of nodes that are engaging in choices to disrupt the protocol should be bound by their stake weight to the denial of service.  Two options exits for nodes:
+
+* a node can outrun previous nodes in virtual generation and submit a concurrent fork
+* a node can withold a vote to observe multiple forks before voting
+
+In both cases, the nodes in the network have several forks to pick from concurrently, even though each fork represents a different height.
+
+### Greedy Choice for Concurrent Forks
+
+When evaluating multiple forks, each node should pick the fork that will maximize economic finality, or the latest fork if all are equal.  Economic finality can be measured in terms of lockout on a fork.
+ 
+## Simulation
+
+A test library function exists for configuring networks.
```

Comment:
How about adding a link to the file here?  and then moving all this code documentation into that file?